File: python/src/chunkshop/bakeoff/output.py
"""Output writers for bakeoff runs (multi-backend).

Three files land in `out_dir`:
- `results.json` — raw BakeoffResults, round-trips via pydantic.
- `report.md` — human-readable side-by-side leaderboard + per-backend
  leaderboards + per-query detail + statistical-power note.
- `recommended.yaml` — the top-MRR combo rendered as a runnable chunkshop
  `CellConfig`. Ties broken by preferring postgres for the recommended
  emit (other backends should match accuracy when chunks + vectors do).
"""
from __future__ import annotations

from pathlib import Path

import yaml

from chunkshop.bakeoff.config import BakeoffConfig, BakeoffResults, ComboResult
from chunkshop.bakeoff.keys import chunker_key, embedder_key, target_display_keys


_BACKEND_ORDER = ["postgres", "mariadb", "sqlite", "clickhouse"]
_BACKEND_SHORT = {"postgres": "PG", "mariadb": "MD", "sqlite": "SQ", "clickhouse": "CH"}


def _combo_target_key(combo: ComboResult) -> str:
    return combo.target_key or combo.backend


def _target_short(key: str) -> str:
    if key == "postgres_cosine":
        return "PG cos"
    if key == "postgres_inner_product":
        return "PG ip"
    if key == "postgres_l2":
        return "PG l2"
    return _BACKEND_SHORT.get(key, key[:2].upper())


def write_results_json(results: BakeoffResults, out_dir: Path) -> Path:
    """Dump raw BakeoffResults to `{out_dir}/results.json`."""
    out = out_dir / "results.json"
    out.write_text(results.model_dump_json(indent=2))
    return out


def _targets_present(results: BakeoffResults) -> list[str]:
    """Return target keys that appear in `results.combos`, preserving result order."""
    seen = []
    seen_set = set()
    for combo in results.combos:
        key = _combo_target_key(combo)
        if key not in seen_set:
            seen.append(key)
            seen_set.add(key)
    ordered = [b for b in _BACKEND_ORDER if b in seen_set]
    ordered += [b for b in seen if b not in ordered]
    extras = sorted(seen_set - set(ordered))
    return ordered + extras


def _combos_for_target(results: BakeoffResults, target_key: str) -> list[ComboResult]:
    return [c for c in results.combos if _combo_target_key(c) == target_key]


def _per_backend_leaderboard_lines(
    cfg: BakeoffConfig, combos: list[ComboResult], target_key: str
) -> list[str]:
    if not combos:
        return [f"_(no combos for {target_key})_", ""]
    ranked = sorted(combos, key=lambda c: -c.aggregate.get("mrr", 0))
    header_cols = " | ".join(f"r@{k}" for k in cfg.scoring.k)
    sep_cells = "|".join(["---"] * (len(cfg.scoring.k) + 8))
    lines = [
        f"| # | Chunker | Embedder | {header_cols} | MRR | chunks | ingest_s | embed_s | query_s |",
        f"|{sep_cells}|",
    ]
    for i, c in enumerate(ranked, start=1):
        rk = [f"{c.aggregate.get(f'recall_at_{k}', 0):.3f}" for k in cfg.scoring.k]
        mrr = f"{c.aggregate.get('mrr', 0):.3f}"
        lines.append(
            f"| {i} | `{c.chunker_label}` | `{c.embedder_label}` | "
            + " | ".join(rk)
            + f" | {mrr} | {c.ingest_chunks} | {c.ingest_wall_seconds:.2f} | "
            f"{c.ingest_embed_seconds:.2f} | {c.query_wall_seconds:.3f} |"
        )
    return lines


def _cross_backend_table_lines(
    results: BakeoffResults, target_keys: list[str]
) -> list[str]:
    """One row per (chunker, embedder) combo. Columns: per-target MRR + ingest_s + query_s."""
    # Build (chunker_key, embedder_key) -> {target_key: combo} index
    index: dict[tuple[str, str], dict[str, ComboResult]] = {}
    for c in results.combos:
        key = (c.chunker_key, c.embedder_key)
        index.setdefault(key, {})[_combo_target_key(c)] = c
    if not index:
        return ["_(no combos)_", ""]

    # Stable ordering: pick a canonical combo (any) for label; sort by best MRR across backends desc.
    def _best_mrr(by_backend: dict[str, ComboResult]) -> float:
        return max((c.aggregate.get("mrr", 0) for c in by_backend.values()), default=0)

    ordered_keys = sorted(
        index.keys(), key=lambda k: -_best_mrr(index[k])
    )

    short = [_target_short(b) for b in target_keys]
    header = (
        ["Chunker", "Embedder"]
        + [f"{s} MRR" for s in short]
        + [f"{s} ing_s" for s in short]
        + [f"{s} qry_s" for s in short]
    )
    sep = "|".join(["---"] * len(header))
    lines = [f"| {' | '.join(header)} |", f"|{sep}|"]
    for key in ordered_keys:
        by_b = index[key]
        sample = next(iter(by_b.values()))
        cells = [f"`{sample.chunker_label}`", f"`{sample.embedder_label}`"]
        for b in target_keys:
            c = by_b.get(b)
            cells.append(f"{c.aggregate.get('mrr', 0):.3f}" if c else "—")
        for b in target_keys:
            c = by_b.get(b)
            cells.append(f"{c.ingest_wall_seconds:.2f}" if c else "—")
        for b in target_keys:
            c = by_b.get(b)
            cells.append(f"{c.query_wall_seconds:.3f}" if c else "—")
        lines.append(f"| {' | '.join(cells)} |")
    return lines


def write_report_md(
    cfg: BakeoffConfig, results: BakeoffResults, out_dir: Path
) -> Path:
    """Render side-by-side multi-target report."""
    target_keys = _targets_present(results)
    n_targets = max(len(target_keys), 1)
    combos_per_target = results.n_combos // n_targets

    lines: list[str] = [
        f"# Bakeoff report: {results.run_name} (multi-target)",
        "",
        f"- Run: {results.started_at}",
        f"- Corpus: {results.corpus_label}",
        f"- Queries: {results.n_queries}",
        f"- Targets: {', '.join(target_keys) if target_keys else '(none)'}",
        f"- Combos per target: {combos_per_target}",
        f"- Total cells: {results.n_combos}",
        "",
        "## Cross-target comparison",
        "",
        "For each (chunker, embedder), MRR + ingest/query wall time per target. "
        "When the targets differ only by Postgres `vector_metric`, chunks and "
        "vectors are identical and MRR changes isolate the pgvector operator.",
        "",
    ]
    lines += _cross_backend_table_lines(results, target_keys)

    # Per-target leaderboards
    for target_key in target_keys:
        combos = _combos_for_target(results, target_key)
        lines += [
            "",
            f"## {target_key} leaderboard",
            "",
        ]
        lines += _per_backend_leaderboard_lines(cfg, combos, target_key)

    # Per-query detail across targets
    lines += [
        "",
        "## Per-query detail (top-1 hit per target × combo)",
        "",
        "| Target | Backend | Chunker | Embedder | Query | Gold | Top-1 | MRR |",
        "|---|---|---|---|---|---|---|---|",
    ]
    # Same ordering as the per-target leaderboard: best MRR first within target.
    for target_key in target_keys:
        combos = sorted(
            _combos_for_target(results, target_key),
            key=lambda c: -c.aggregate.get("mrr", 0),
        )
        for c in combos:
            for pq in c.per_query:
                top1 = pq["top_k"][0]["doc_id"] if pq.get("top_k") else "-"
                lines.append(
                    f"| {target_key} | {c.backend} | `{c.chunker_label}` | `{c.embedder_label}` | "
                    f"{pq['query']} | `{pq['gold_doc_id']}` | "
                    f"`{top1}` | {pq.get('mrr', 0):.3f} |"
                )

    # Query-time embedding cost (per-embedder, unchanged shape)
    if results.query_embed_seconds_by_embedder:
        n = max(results.n_queries, 1)
        lines += [
            "",
            "## Query-time embedding cost",
            "",
            f"Wall time to embed all {results.n_queries} gold queries, per "
            "unique embedder. At production scale this scales by your "
            "expected QPS — useful for choosing between a slower-but-better "
            "embedder and a faster-but-worse one.",
            "",
            "| Embedder | total_s | per_query_ms |",
            "|---|---|---|",
        ]
        for k, total in sorted(
            results.query_embed_seconds_by_embedder.items(), key=lambda kv: kv[1]
        ):
            per_q_ms = (total / n) * 1000.0
            lines.append(f"| `{k}` | {total:.3f} | {per_q_ms:.1f} |")

    # Honesty note scaled to n_queries.
    n = max(results.n_queries, 1)
    lines += [
        "",
        "## Statistical power",
        "",
        f"{results.n_queries} queries means one query flipping moves aggregate recall by "
        f"{1 / n:.3f}. Combos within ~{2 / n:.2f} of the leader are not reliably "
        "distinguishable. Re-run with more queries or a larger corpus before treating "
        "the leaderboard as a tournament result.",
        "",
    ]
    out = out_dir / "report.md"
    out.write_text("\n".join(lines))
    return out


def _pick_recommended(results: BakeoffResults) -> ComboResult:
    """Top MRR combo, ties broken by preferring postgres > mariadb > sqlite."""
    backend_pref = {b: i for i, b in enumerate(_BACKEND_ORDER)}

    def sort_key(c: ComboResult):
        return (
            -c.aggregate.get("mrr", 0),
            backend_pref.get(c.backend, 99),
            c.chunker_key,
            c.embedder_key,
        )

    return sorted(results.combos, key=sort_key)[0]


def write_recommended_yaml(
    cfg: BakeoffConfig, results: BakeoffResults, out_dir: Path
) -> Path:
    """Render the top-MRR combo as a runnable CellConfig YAML.

    Multi-backend run: ties broken by backend preference (postgres > mariadb >
    sqlite). The recommended target inherits the winner's backend type; the
    user can swap it after copying the file.
    """
    top = _pick_recommended(results)

    winner_chunker = next(
        c for c in cfg.matrix.chunkers if chunker_key(c) == top.chunker_key
    )
    winner_embedder = next(
        e for e in cfg.matrix.embedders if embedder_key(e) == top.embedder_key
    )
    target_keys = target_display_keys(cfg.targets)
    top_target_key = _combo_target_key(top)
    winner_target = next(
        t for key, t in zip(target_keys, cfg.targets)
        if key == top_target_key and t.type == top.backend
    )

    recommended = {
        "# NOTE": (
            f"Top combo from multi-target bakeoff '{results.run_name}' "
            f"(target={top_target_key}, backend={top.backend}, "
            f"MRR={top.aggregate.get('mrr', 0):.3f}, "
            f"r@1={top.aggregate.get('recall_at_1', 0):.3f}). "
            "Point `source` at your real corpus before running `chunkshop ingest`."
        ),
        "cell_name": f"{results.run_name}_recommended",
        "source": cfg.source.model_dump(exclude_none=True, by_alias=True),
        "framer": cfg.framer.model_dump(exclude_none=True, by_alias=True),
        "chunker": winner_chunker.model_dump(exclude_none=True, by_alias=True),
        "embedder": winner_embedder.model_dump(exclude_none=True, by_alias=True),
        "target": {
            "type": top.backend,
            "dsn_env": winner_target.dsn_env,
            "database": winner_target.database_name,
            "table": f"{results.run_name}_production",
            "mode": "overwrite",
            **(
                {"vector_metric": winner_target.vector_metric}
                if top.backend == "postgres"
                else {}
            ),
        },
    }
    out = out_dir / "recommended.yaml"
    out.write_text(yaml.safe_dump(recommended, sort_keys=False))
    return out

File: rust/chunkshop/src/bakeoff/output.rs
//! Output writers. Mirrors `python/src/chunkshop/bakeoff/output.py`.
//!
//! Three files land in `out_dir`:
//! - `results.json` — raw `BakeoffResults` as JSON
//! - `report.md`    — leaderboard + per-query detail + statistical-power note
//! - `recommended.yaml` — top-MRR combo as a runnable `CellConfig` YAML
//!
//! report.md formatting is deliberately byte-comparable with Python's so the
//! cross-language parity test can diff the leaderboard tables directly.

use std::path::{Path, PathBuf};

use anyhow::{Context, Result};
use serde_json::{json, Value};

use super::config::{BakeoffConfig, BakeoffResults, ComboResult};
use super::keys::{chunker_key, embedder_key};

pub fn write_results_json(results: &BakeoffResults, out_dir: &Path) -> Result<PathBuf> {
    let out = out_dir.join("results.json");
    let text = serde_json::to_string_pretty(results)?;
    std::fs::write(&out, text).with_context(|| format!("write {}", out.display()))?;
    Ok(out)
}

fn fmt_f3(x: f64) -> String {
    format!("{:.3}", x)
}

pub fn write_report_md(
    cfg: &BakeoffConfig,
    results: &BakeoffResults,
    out_dir: &Path,
) -> Result<PathBuf> {
    // Sort by MRR desc. Stable sort so Python's input order breaks ties.
    let mut ranked: Vec<&ComboResult> = results.combos.iter().collect();
    ranked.sort_by(|a, b| {
        let am = a.aggregate.get("mrr").copied().unwrap_or(0.0);
        let bm = b.aggregate.get("mrr").copied().unwrap_or(0.0);
        bm.partial_cmp(&am).unwrap_or(std::cmp::Ordering::Equal)
    });

    let header_cols = cfg
        .scoring
        .k
        .iter()
        .map(|k| format!("r@{k}"))
        .collect::<Vec<_>>()
        .join(" | ");
    // Columns: # | Chunker | Embedder | r@k... | MRR | chunks | ingest_s | embed_s
    let sep_cells = std::iter::repeat("---")
        .take(cfg.scoring.k.len() + 7)
        .collect::<Vec<_>>()
        .join("|");

    let mut lines: Vec<String> = vec![
        format!("# Bakeoff report: {}", results.run_name),
        String::new(),
        format!("- Run: {}", results.started_at),
        format!("- Corpus: {}", results.corpus_label),
        format!("- Queries: {}", results.n_queries),
        format!("- Combos: {}", results.n_combos),
        String::new(),
        "## Leaderboard (sorted by MRR)".into(),
        String::new(),
        format!("| # | Chunker | Embedder | {header_cols} | MRR | chunks | ingest_s | embed_s |"),
        format!("|{sep_cells}|"),
    ];
    for (i, c) in ranked.iter().enumerate() {
        let rk: Vec<String> = cfg
            .scoring
            .k
            .iter()
            .map(|k| {
                fmt_f3(
                    c.aggregate
                        .get(&format!("recall_at_{k}"))
                        .copied()
                        .unwrap_or(0.0),
                )
            })
            .collect();
        let mrr = fmt_f3(c.aggregate.get("mrr").copied().unwrap_or(0.0));
        lines.push(format!(
            "| {n} | `{chunker}` | `{embedder}` | {rks} | {mrr} | {chunks} | {ingest:.2} | {embed:.2} |",
            n = i + 1,
            chunker = c.chunker_label,
            embedder = c.embedder_label,
            rks = rk.join(" | "),
            mrr = mrr,
            chunks = c.ingest_chunks,
            ingest = c.ingest_wall_seconds,
            embed = c.ingest_embed_seconds,
        ));
    }

    lines.push(String::new());
    lines.push("## Per-query detail (top-1 hit per combo)".into());
    lines.push(String::new());
    lines.push("| Chunker | Embedder | Query | Gold | Top-1 | MRR |".into());
    lines.push("|---|---|---|---|---|---|".into());
    for c in &ranked {
        for pq in &c.per_query {
            let top1 = pq
                .top_k
                .first()
                .map(|h| h.doc_id.as_str())
                .unwrap_or("-");
            let mrr = fmt_f3(pq.scores.get("mrr").copied().unwrap_or(0.0));
            lines.push(format!(
                "| `{chunker}` | `{embedder}` | {query} | `{gold}` | `{top1}` | {mrr} |",
                chunker = c.chunker_label,
                embedder = c.embedder_label,
                query = pq.query,
                gold = pq.gold_doc_id,
            ));
        }
    }

    let n = results.n_queries.max(1) as f64;

    // Query-time embedding cost: per-embedder wall time during scoring.
    if !results.query_embed_seconds_by_embedder.is_empty() {
        lines.push(String::new());
        lines.push("## Query-time embedding cost".into());
        lines.push(String::new());
        lines.push(format!(
            "Wall time to embed all {} gold queries, per unique embedder. \
             At production scale this scales by your expected QPS — useful \
             for choosing between a slower-but-better embedder and a \
             faster-but-worse one.",
            results.n_queries
        ));
        lines.push(String::new());
        lines.push("| Embedder | total_s | per_query_ms |".into());
        lines.push("|---|---|---|".into());
        // Sort by total time ascending (fastest first).
        let mut entries: Vec<(&String, &f64)> = results
            .query_embed_seconds_by_embedder
            .iter()
            .collect();
        entries.sort_by(|a, b| a.1.partial_cmp(b.1).unwrap_or(std::cmp::Ordering::Equal));
        for (k, total) in entries {
            let per_q_ms = (total / n) * 1000.0;
            lines.push(format!("| `{k}` | {total:.3} | {per_q_ms:.1} |"));
        }
    }

    lines.push(String::new());
    lines.push("## Statistical power".into());
    lines.push(String::new());
    lines.push(format!(
        "{} queries means one query flipping moves aggregate recall by \
         {:.3}. Combos within ~{:.2} of the leader are not reliably \
         distinguishable. Re-run with more queries or a larger corpus before \
         treating the leaderboard as a tournament result.",
        results.n_queries,
        1.0 / n,
        2.0 / n
    ));
    lines.push(String::new());

    let out = out_dir.join("report.md");
    std::fs::write(&out, lines.join("\n")).with_context(|| format!("write {}", out.display()))?;
    Ok(out)
}

/// Render the top-MRR combo as a runnable CellConfig YAML.
///
/// Functional equivalence with Python's emission, not byte-equality:
/// - Same chunker config + same embedder config + same source/framer.
/// - cell_name suffix `_recommended` matches Python.
/// - Default mode = `overwrite`, schema = bakeoff schema, table = `{run_name}_production`.
pub fn write_recommended_yaml(
    cfg: &BakeoffConfig,
    results: &BakeoffResults,
    out_dir: &Path,
) -> Result<PathBuf> {
    let mut ranked: Vec<&ComboResult> = results.combos.iter().collect();
    ranked.sort_by(|a, b| {
        let am = a.aggregate.get("mrr").copied().unwrap_or(0.0);
        let bm = b.aggregate.get("mrr").copied().unwrap_or(0.0);
        bm.partial_cmp(&am).unwrap_or(std::cmp::Ordering::Equal)
    });
    let top = ranked
        .first()
        .ok_or_else(|| anyhow::anyhow!("no combos to recommend"))?;

    let winner_chunker = cfg
        .matrix
        .chunkers
        .iter()
        .find(|c| chunker_key(c).map(|k| k == top.chunker_key).unwrap_or(false))
        .ok_or_else(|| anyhow::anyhow!("winner chunker not found in matrix"))?;
    let winner_embedder = cfg
        .matrix
        .embedders
        .iter()
        .find(|e| embedder_key(e) == top.embedder_key)
        .ok_or_else(|| anyhow::anyhow!("winner embedder not found in matrix"))?;

    // Round-trip through serde_json::Value so we can drop fields the Python
    // version doesn't emit. Embedder/chunker/source/framer don't impl Serialize
    // (they're Deserialize-only), so we re-render them by re-serializing
    // through the matching shape. For simplicity and parity with Python, we
    // hand-build minimal maps from the discriminator + known fields.
    let chunker_yaml = chunker_to_yaml_value(winner_chunker)?;
    let embedder_yaml = embedder_to_yaml_value(winner_embedder);
    let source_yaml = source_to_yaml_value(&cfg.source);
    let framer_yaml = framer_to_yaml_value(
        cfg.framer
            .as_ref()
            .unwrap_or(&crate::config::FramerConfig::Identity(
                crate::config::IdentityFramerConfig {},
            )),
    );

    let note = format!(
        "Top combo from bakeoff '{}' (MRR={:.3}, r@1={:.3}). Point `source` \
         at your real corpus before running `chunkshop ingest`.",
        results.run_name,
        top.aggregate.get("mrr").copied().unwrap_or(0.0),
        top.aggregate.get("recall_at_1").copied().unwrap_or(0.0),
    );

    let recommended = json!({
        "# NOTE": note,
        "cell_name": format!("{}_recommended", results.run_name),
        "source": source_yaml,
        "framer": framer_yaml,
        "chunker": chunker_yaml,
        "embedder": embedder_yaml,
        "target": {
            "dsn_env": cfg.target.dsn_env,
            "schema": cfg.target.schema_name,
            "table": format!("{}_production", results.run_name),
            "mode": "overwrite",
        },
    });

    let yaml_text = serde_yml::to_string(&recommended)?;
    let out = out_dir.join("recommended.yaml");
    std::fs::write(&out, yaml_text).with_context(|| format!("write {}", out.display()))?;
    Ok(out)
}

fn chunker_to_yaml_value(c: &crate::config::ChunkerConfig) -> Result<Value> {
    use crate::config::ChunkerConfig as C;
    Ok(match c {
        C::Hierarchy(c) => json!({
            "type": "hierarchy",
            "prefix_heading": c.prefix_heading,
            "min_section_chars": c.min_section_chars,
            "max_chars": c.max_chars,
        }),
        C::SentenceAware(c) => json!({
            "type": "sentence_aware",
            "doc_type": c.doc_type,
            "max_chars": c.max_chars,
            "min_chars": c.min_chars,
        }),
        C::FixedOverlap(c) => json!({
            "type": "fixed_overlap",
            "window_words": c.window_words,
            "step_words": c.step_words,
        }),
        C::NeighborExpand(c) => json!({
            "type": "neighbor_expand",
            "base": chunker_to_yaml_value(&c.base)?,
            "window": c.window,
        }),
        C::Semantic(_) | C::SummaryEmbed(_) | C::HierarchicalSummary(_) => {
            return Err(anyhow::anyhow!(
                "recommended.yaml emission for this chunker variant is not implemented; \
                 these are out of the bakeoff matrix today."
            ));
        }
    })
}

fn embedder_to_yaml_value(e: &crate::config::FastembedEmbedderConfig) -> Value {
    let mut m = serde_json::Map::new();
    m.insert("type".into(), json!("fastembed"));
    m.insert("model_name".into(), json!(e.model_name));
    m.insert("dim".into(), json!(e.dim));
    m.insert("batch_size".into(), json!(e.batch_size));
    if let Some(t) = e.threads {
        m.insert("threads".into(), json!(t));
    }
    Value::Object(m)
}

fn source_to_yaml_value(s: &crate::config::SourceConfig) -> Value {
    use crate::config::SourceConfig as S;
    match s {
        S::Files(f) => json!({
            "type": "files",
            "glob": f.glob,
            "id_from": f.id_from,
            "encoding": f.encoding,
        }),
        S::JsonCorpus(j) => json!({
            "type": "json_corpus",
            "path": j.path,
            "documents_key": j.documents_key,
            "id_field": j.id_field,
            "content_field": j.content_field,
            "title_field": j.title_field,
        }),
        S::PgTable(p) => {
            let mut m = serde_json::Map::new();
            m.insert("type".into(), json!("pg_table"));
            m.insert("dsn_env".into(), json!(p.dsn_env));
            m.insert("schema".into(), json!(p.schema_name));
            m.insert("table".into(), json!(p.table));
            m.insert("id_column".into(), json!(p.id_column));
            m.insert("content_column".into(), json!(p.content_column));
            if let Some(t) = &p.title_column {
                m.insert("title_column".into(), json!(t));
            }
            if let Some(w) = &p.where_clause {
                m.insert("where".into(), json!(w));
            }
            Value::Object(m)
        }
        S::Http(h) => {
            let mut m = serde_json::Map::new();
            m.insert("type".into(), json!("http"));
            m.insert("urls".into(), json!(h.urls));
            if let Some(s) = &h.sitemap {
                m.insert("sitemap".into(), json!(s));
            }
            Value::Object(m)
        }
        S::S3(s3) => {
            let mut m = serde_json::Map::new();
            m.insert("type".into(), json!("s3"));
            m.insert("bucket".into(), json!(s3.bucket));
            m.insert("prefix".into(), json!(s3.prefix));
            if let Some(e) = &s3.endpoint_url {
                m.insert("endpoint_url".into(), json!(e));
            }
            Value::Object(m)
        }
        S::ClickhouseTable(c) => {
            let mut m = serde_json::Map::new();
            m.insert("type".into(), json!("clickhouse_table"));
            m.insert("dsn_env".into(), json!(c.dsn_env));
            m.insert("database".into(), json!(c.database_name));
            m.insert("table".into(), json!(c.table));
            m.insert("id_column".into(), json!(c.id_column));
            m.insert("content_column".into(), json!(c.content_column));
            if let Some(t) = &c.title_column {
                m.insert("title_column".into(), json!(t));
            }
            if let Some(w) = &c.where_clause {
                m.insert("where".into(), json!(w));
            }
            if !c.metadata_columns.is_empty() {
                m.insert("metadata_columns".into(), json!(c.metadata_columns));
            }
            Value::Object(m)
        }
        S::Inline(_) => json!({ "type": "inline" }),
    }
}

fn framer_to_yaml_value(f: &crate::config::FramerConfig) -> Value {
    use crate::config::FramerConfig as F;
    match f {
        F::Identity(_) => json!({ "type": "identity" }),
        F::HeadingBoundary(h) => {
            let mut m = serde_json::Map::new();
            m.insert("type".into(), json!("heading_boundary"));
            m.insert("pattern".into(), json!(h.pattern));
            m.insert("title_from_heading".into(), json!(h.title_from_heading));
            Value::Object(m)
        }
        F::RegexBoundary(r) => {
            let mut m = serde_json::Map::new();
            m.insert("type".into(), json!("regex_boundary"));
            m.insert("split_pattern".into(), json!(r.split_pattern));
            if let Some(p) = &r.title_pattern {
                m.insert("title_pattern".into(), json!(p));
            }
            m.insert("body_starts_with_match".into(), json!(r.body_starts_with_match));
            Value::Object(m)
        }
        F::Jsonpath(j) => {
            let mut m = serde_json::Map::new();
            m.insert("type".into(), json!("jsonpath"));
            m.insert("row_path".into(), json!(j.row_path));
            m.insert("body_path".into(), json!(j.body_path));
            if let Some(t) = &j.title_path {
                m.insert("title_path".into(), json!(t));
            }
            Value::Object(m)
        }
    }
}
